src/go/serviceconfig/service_config_fetcher.go

```diff
@@ -0,0 +1,237 @@
+	if scf.mf == nil && *flags.ServiceAccountKey == "" {
+		return "", 0, fmt.Errorf("If --non_gcp is specified, --service_account_key has to be specified.")
+	}
+	if *flags.ServiceAccountKey != "" {
+		return util.GenerateAccessTokenFromFile(*flags.ServiceAccountKey)
+	}
+	return scf.mf.FetchAccessToken()
+}
+
+// TODO(jcwang) cleanup here. This function is redundant.
+func (scf *ServiceConfigFetcher) fetchRollouts() (*smpb.ListServiceRolloutsResponse, error) {
+	token, _, err := scf.accessToken()
+	if err != nil {
+		return nil, fmt.Errorf("fail to get access token: %v", err)
+	}
+
+	return scf.callServiceManagementRollouts(util.FetchRolloutsURL(*flags.ServiceManagementURL, scf.serviceName), token)
+}
+
+func (scf *ServiceConfigFetcher) callServiceManagementRollouts(path, token string) (*smpb.ListServiceRolloutsResponse, error) {
+	var err error
+	var resp *http.Response
+	if resp, err = scf.callWithAccessToken(path, token); err != nil {
+		return nil, err
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("fail to read response body: %s", err)
+	}
+	defer resp.Body.Close()
+	rolloutsResponse := new(smpb.ListServiceRolloutsResponse)
+	if err := proto.Unmarshal(body, rolloutsResponse); err != nil {
+		return nil, fmt.Errorf("fail to unmarshal ListServiceRolloutsResponse: %s", err)
+	}
+	return rolloutsResponse, nil
+}
+
+func (scf *ServiceConfigFetcher) callServiceManagement(path, token string) (*confpb.Service, error) {
+	var err error
+	var resp *http.Response
+	if resp, err = scf.callWithAccessToken(path, token); err != nil {
+		return nil, err
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("fail to read response body: %s", err)
+	}
+	defer resp.Body.Close()
+	service := new(confpb.Service)
+	if err := proto.Unmarshal(body, service); err != nil {
+		return nil, fmt.Errorf("fail to unmarshal Service: %v", err)
+	}
+	return service, nil
+}
+
+func (scf *ServiceConfigFetcher) callWithAccessToken(path, token string) (*http.Response, error) {
+	req, _ := http.NewRequest("GET", path, nil)
+	req.Header.Add("Authorization", "Bearer "+token)
+	req.Header.Set("Content-Type", "application/x-protobuf")
+	resp, err := scf.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		resp.Body.Close()
+		return nil, fmt.Errorf("http call to %s returns not 200 OK: %v", path, resp.Status)
+	}
+	return resp, nil
+}
+
+// Fetch the service config by given configId. If configId is empty, try to
+// fetch the latest service config,.
+func (scf *ServiceConfigFetcher) FetchConfig(configId string) (*confpb.Service, error) {
+	if configId != "" {
+		scf.curConfigId = configId
+		token, _, err := scf.accessToken()
+		if err != nil {
+			return nil, fmt.Errorf("fail to get access token: %v", err)
+		}
+		return scf.callServiceManagement(util.FetchConfigURL(*flags.ServiceManagementURL, scf.serviceName, configId), token)
```

Comment:
I recommend removing all the flags here, change them to options now that ServiceConfigFetcher is a struct. Otherwise it's hard to read; I don't even know where this flag is defined.